### PR TITLE
Feature puma patches [8/x] Batch processing of tree ref_coords [2/x] Linear geometries

### PR DESCRIPTION
**Pull request by sandro-elsweijer**
**_Describe your changes here:_**
Add batch processing for linear geometries

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] New Datatypes are added to t8indent_custom_datatypes.txt
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_geometry/t8_geometry_helpers.c [hunk 1]
```diff
@@ -26,134 +26,145 @@
...
   /* The algorithm is able to calculate any point in a triangle or tetrahedron using barycentric coordinates.
    * All points are calculated by the sum of each corner point (e.g. p1 -> corner point 1) multiplied by a
    * scalar, which in this case are the reference coordinates (ref_coords).
    */
   double temp[3] = { 0 };
 
-  for (int i = 0; i < corner_value_dim; i++) {
-    temp[i]
-      = (corner_values[corner_value_dim + i] - corner_values[i]) * coefficients[0] /* (p2 - p1) * ref_coords */
+  for (int dim = 0; dim < corner_value_dim; dim++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:66]:
```suggestion
  for (int idim = 0; idim < corner_value_dim; idim++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 2]
```diff
@@ -26,134 +26,145 @@
...
 
 void
 t8_geom_compute_linear_geometry (t8_eclass_t tree_class, const double *tree_vertices, const double *ref_coords,
-                                 double out_coords[3])
+                                 const size_t num_coords, double *out_coords)
 {
-  int i;
-  int dimension = t8_eclass_to_dimension[tree_class];
+  int dim;
+  size_t coord;
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:86]:
```suggestion
  size_t icoord;
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 3]
```diff
@@ -26,134 +26,145 @@
...
     /* A vertex has exactly one corner, and we already know its coordinates, since they are
      * the same as the trees coordinates. */
-    for (i = 0; i < 3; i++) {
-      out_coords[i] = tree_vertices[i];
-    }
-    break;
-  case T8_ECLASS_LINE:
-    for (i = 0; i < 3; i++) {
-      out_coords[i] = (tree_vertices[3 + i] - tree_vertices[i]) * ref_coords[0] + tree_vertices[i];
+    for (coord = 0; coord < num_coords; coord++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:93]:
```suggestion
    for (icoord = 0; icoord < num_coords; icoord++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 4]
```diff
@@ -26,134 +26,145 @@
...
+      const size_t offset_domain_dim = coord * T8_ECLASS_MAX_DIM;
+      for (dim = 0; dim < T8_ECLASS_MAX_DIM; dim++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:95]:
```suggestion
      for (idim= 0; idim < T8_ECLASS_MAX_DIM; idim++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 5]
```diff
@@ -26,134 +26,145 @@
...
+        out_coords[offset_domain_dim + dim] = tree_vertices[offset_domain_dim + dim];
+      }
     }
     break;
   case T8_ECLASS_TRIANGLE:
   case T8_ECLASS_TET:
-    t8_geom_triangular_interpolation (ref_coords, tree_vertices, 3, dimension, out_coords);
+    for (coord = 0; coord < num_coords; coord++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:102]:
```suggestion
    for (icoord = 0; icoord < num_coords; icoord++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 6]
```diff
@@ -26,134 +26,145 @@
...
-    /* Get a triangle at the specific height */
     double tri_vertices[9];
-    for (i = 0; i < 9; i++) {
-      tri_vertices[i] = (tree_vertices[9 + i] - tree_vertices[i]) * ref_coords[2] + tree_vertices[i];
-    }
-    for (i = 0; i < 3; i++) {
-      out_coords[i] = (tri_vertices[3 + i] - tri_vertices[i]) * ref_coords[0]
-                      + (tri_vertices[6 + i] - tri_vertices[3 + i]) * ref_coords[1] + tri_vertices[i];
+    double line_vertices[6];
+    for (coord = 0; coord < num_coords; coord++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:112]:
```suggestion
    for (icoord = 0; icoord < num_coords; icoord++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 7]
```diff
@@ -26,134 +26,145 @@
...
+      const size_t offset_tree_dim = coord * dimension;
+      const size_t offset_domain_dim = coord * T8_ECLASS_MAX_DIM;
+      /* Prisminterpolation, via height and triangle */
+      /* Get a triangle at the specific height */
+      for (int tri_vertex = 0; tri_vertex < 3; tri_vertex++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:117]:
```suggestion
      for (int itri_vertex = 0; itri_vertex < 3; itri_vertex++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 8]
```diff
@@ -26,134 +26,145 @@
...
+      }
+      t8_geom_triangular_interpolation (ref_coords + offset_tree_dim, tri_vertices, T8_ECLASS_MAX_DIM, 2,
+                                        out_coords + offset_domain_dim);
     }
   } break;
+  case T8_ECLASS_LINE:
   case T8_ECLASS_QUAD:
   case T8_ECLASS_HEX:
-    t8_geom_linear_interpolation (ref_coords, tree_vertices, 3, dimension, out_coords);
+    for (coord = 0; coord < num_coords; coord++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:132]:
```suggestion
    for (icoord = 0; icoord < num_coords; icoord++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 9]
```diff
@@ -26,134 +26,145 @@
...
-    double ray[3], lambda, quad_coords[3], length, length2;
-    length = 0;
-    length2 = 0;
-    quad_coords[2] = 0;
-    /*vertices = tree_vertices */
-    /*coordinates = out_coords */
-    /*vertex_coords = len * corner_coords = ref_coords */
+    double base_coords[2];
+    double vec[3];
+    for (coord = 0; coord < num_coords; coord++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:142]:
```suggestion
    for (icoord = 0; icoord < num_coords; icoord++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 10]
```diff
@@ -26,134 +26,145 @@
...
+       * interpolation on the base and then we interpolate via the height to the top vertex  */
 
-    /*In this case, the vertex is the tip of the parent pyramid and we don't have to compute
-       * anything.*/
-    if (ref_coords[0] == 1. && ref_coords[1] == 1. && ref_coords[2] == 1.) {
-      for (i = 0; i < 3; i++) {
-        out_coords[i] = tree_vertices[12 + i];
+      /* Project point on base */
+      if (ref_coords[offset_tree_dim + 2] != 1.) {
+        for (dim = 0; dim < 2; dim++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:150]:
```suggestion
        for (idim = 0; idim < 2; idim++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 11]
```diff
@@ -26,134 +26,145 @@
...
+          base_coords[dim] = 1 - (1 - ref_coords[offset_tree_dim + dim]) / (1 - ref_coords[offset_tree_dim + 2]);
+        }
+      }
+      else {
+        for (dim = 0; dim < T8_ECLASS_MAX_DIM; dim++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:155]:
```suggestion
        for (idim = 0; idim < T8_ECLASS_MAX_DIM; idim++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 12]
```diff
@@ -26,134 +26,145 @@
...
+          out_coords[offset_domain_dim + dim] = tree_vertices[4 * T8_ECLASS_MAX_DIM + dim];
+        }
+        continue;
+      }
+      /* Get a quad interpolation of the base */
+      t8_geom_linear_interpolation (base_coords, tree_vertices, T8_ECLASS_MAX_DIM, 2, out_coords + offset_domain_dim);
+      /* Get vector from base to pyramid tip */
+      t8_vec_diff (tree_vertices + 4 * T8_ECLASS_MAX_DIM, out_coords + offset_domain_dim, vec);
+      /* Add vector to base */
+      for (dim = 0; dim < 3; dim++) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:165]:
```suggestion
      for (idim = 0; idim < 3; idim++) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 13]
```diff
@@ -176,8 +187,8 @@ t8_geom_compute_linear_axis_aligned_geometry (t8_eclass_t tree_class, const doub
     /* The two vertices of a line must have two matching coordinates to be
      * axis-aligned. A quad needs one matching coordinate. */
     int n_equal_coords = 0;
-    for (int dim = 0; dim < 3; ++dim) {
-      if (abs (tree_vertices[dim] - tree_vertices[3 + dim]) <= SC_EPS) {
+    for (int dim = 0; dim < T8_ECLASS_MAX_DIM; ++dim) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:190]:
```suggestion
    for (int idim = 0; idim < T8_ECLASS_MAX_DIM; ++idim) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 14]
```diff
@@ -189,15 +200,19 @@ t8_geom_compute_linear_axis_aligned_geometry (t8_eclass_t tree_class, const doub
...
   /* Compute vector between both points */
   double vector[3];
-  t8_vec_diff (tree_vertices + 3, tree_vertices, vector);
+  t8_vec_diff (tree_vertices + T8_ECLASS_MAX_DIM, tree_vertices, vector);
 
   /* Compute the coordinates of the reference point. */
-  for (int dim = 0; dim < 3; ++dim) {
-    out_coords[dim] = tree_vertices[dim];
-    out_coords[dim] += ref_coords[dim] * vector[dim];
+  for (size_t coord = 0; coord < num_coords; ++coord) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:209]:
```suggestion
  for (size_t icoord = 0; icoord < num_coords; ++icoord) {
```

File: src/t8_geometry/t8_geometry_helpers.c [hunk 15]
```diff
@@ -189,15 +200,19 @@ t8_geom_compute_linear_axis_aligned_geometry (t8_eclass_t tree_class, const doub
...
+    const size_t offset_tree_dim = coord * dimension;
+    const size_t offset_domain_dim = coord * T8_ECLASS_MAX_DIM;
+    for (int dim = 0; dim < T8_ECLASS_MAX_DIM; ++dim) {
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_helpers.c:212]:
```suggestion
    for (int idim = 0; idim < T8_ECLASS_MAX_DIM; ++idim) {
```

**Comment by sandro-elsweijer** [comment]
> Nice work! Just nit-picker stuff to do.

Good spot, changed that

**Review by jmark — APPROVED** [review]

**Pull request merged**